File: typescript/src/shared/api_response.ts
import { RejectionCode } from "./rejection";
import type { LinkedIdentityType } from "../auth";

/** Raw backend rejection details before bounded codes and methods are decoded. */
export interface ApiRejectedDetailsWire {
  reason: string;
  rejection_code?: string;
  error_code?: string;
  /** Primary method of a deterministic identity owner; unknown values are ignored. */
  existing_method?: string;
  error_log_id?: string;
}

export type ApiResponse<T> =
  | { status: "success"; body: T }
  | { status: "error"; error_details: ApiRejectedDetailsWire };

/** Structured rejection details with correlation and transport context. */
export class ApiRejectedDetails {
  readonly reason: string;
  readonly rejectionCode?: RejectionCode;
  readonly errorCode?: string;
  readonly existingMethod?: LinkedIdentityType;
  readonly errorLogId?: string;
  readonly requestId?: string;
  /**
   * HTTP status of the response that carried this rejection. Set by the HTTP
   * client when the rejection rode a non-2xx response; not present in the
   * backend JSON. Lets callers classify rejections at the transport level
   * (e.g. `isUnauthorized`) without matching on backend error strings.
   */
  readonly httpStatus?: number;

  constructor(params: {
    reason: string;
    rejectionCode?: RejectionCode;
    errorCode?: string;
    existingMethod?: LinkedIdentityType;
    errorLogId?: string;
    requestId?: string;
    httpStatus?: number;
  }) {
    this.reason = params.reason;
    this.rejectionCode = params.rejectionCode;
    this.errorCode = params.errorCode;
    this.existingMethod = params.existingMethod;
    this.errorLogId = params.errorLogId;
    this.requestId = params.requestId;
    this.httpStatus = params.httpStatus;
  }

  static fromWire(
    wire: ApiRejectedDetailsWire,
    requestId?: string,
    httpStatus?: number,
  ): ApiRejectedDetails {
    return new ApiRejectedDetails({
      reason: wire.reason,
      rejectionCode: wire.rejection_code
        ? RejectionCode.from(wire.rejection_code)
        : undefined,
      errorCode: wire.error_code,
      existingMethod: linkedIdentityType(wire.existing_method),
      errorLogId: wire.error_log_id,
      requestId,
      httpStatus,
    });
  }

  toString(): string {
    const lines = [`Reason: ${this.reason}`];
    if (this.rejectionCode) {
      lines.push(`Rejection Code: ${this.rejectionCode.label()}`);
    }
    if (this.errorCode) {
      lines.push(`Error Code: ${this.errorCode}`);
    }
    if (this.existingMethod) {
      lines.push(`Existing Method: ${this.existingMethod}`);
    }
    if (this.errorLogId) {
      lines.push(`Error Log ID: ${this.errorLogId}`);
    }
    if (this.requestId) {
      lines.push(`Request ID: ${this.requestId}`);
    }
    return lines.join("\n");
  }
}

/** Preserve known public method guidance while tolerating future backend variants. */
function linkedIdentityType(
  value: string | undefined,
): LinkedIdentityType | undefined {
  return value === "email" ||
    value === "google" ||
    value === "x" ||
    value === "wallet"
    ? value
    : undefined;
}

export function isApiResponse<T>(value: unknown): value is ApiResponse<T> {
  if (typeof value !== "object" || value === null || !("status" in value)) {
    return false;
  }

  const status = (value as { status?: unknown }).status;
  if (status === "success") {
    return "body" in value;
  }
  if (status === "error") {
    return "error_details" in value;
  }
  return false;
}

File: python/src/lightcone_sdk/shared/api_response.py
"""Generic API response wrapper and structured rejection details."""

from __future__ import annotations

from dataclasses import dataclass, replace
from typing import Generic, Literal, Optional, TypeVar, cast

from .rejection import RejectionCode

T = TypeVar("T")
# Public login-method names exposed in bounded identity conflicts.
LinkedIdentityType = Literal["email", "google", "x", "wallet"]


@dataclass(frozen=True)
class ApiRejectedDetails:
    """Structured backend rejection with bounded method and transport context.

    Unknown future ``existing_method`` values become ``None`` so callers retain
    the stable rejection code and can fall back to generic guidance.
    """

    reason: str
    rejection_code: Optional[RejectionCode] = None
    error_code: Optional[str] = None
    existing_method: Optional[LinkedIdentityType] = None
    error_log_id: Optional[str] = None
    request_id: Optional[str] = None
    # HTTP status of the response that carried this rejection. Set by the HTTP
    # client when the rejection rode a non-2xx response; not present in the
    # backend JSON. Lets callers classify rejections at the transport level
    # (e.g. ``lightcone_sdk.error.is_unauthorized``) without matching on
    # backend error strings.
    http_status: Optional[int] = None

    @staticmethod
    def from_dict(data: dict) -> "ApiRejectedDetails":
        return ApiRejectedDetails(
            reason=str(data.get("reason", "")),
            rejection_code=RejectionCode.from_wire(data.get("rejection_code")),
            error_code=data.get("error_code"),
            existing_method=_linked_identity_type(data.get("existing_method")),
            error_log_id=data.get("error_log_id"),
        )

    def with_request_id(self, request_id: Optional[str]) -> "ApiRejectedDetails":
        return replace(self, request_id=request_id)

    def with_http_status(self, http_status: Optional[int]) -> "ApiRejectedDetails":
        return replace(self, http_status=http_status)

    def __str__(self) -> str:
        lines = [f"Reason: {self.reason}"]
        if self.rejection_code is not None:
            lines.append(f"Rejection Code: {self.rejection_code}")
        if self.error_code is not None:
            lines.append(f"Error Code: {self.error_code}")
        if self.existing_method is not None:
            lines.append(f"Existing Method: {self.existing_method}")
        if self.error_log_id is not None:
            lines.append(f"Error Log ID: {self.error_log_id}")
        if self.request_id is not None:
            lines.append(f"Request ID: {self.request_id}")
        return "\n".join(lines)


def _linked_identity_type(value: object) -> Optional[LinkedIdentityType]:
    """Keep known public method guidance and ignore future backend variants."""
    if value in ("email", "google", "x", "wallet"):
        return cast(LinkedIdentityType, value)
    return None


@dataclass(frozen=True)
class ApiResponse(Generic[T]):
    """Generic backend response wrapper.

    Success shape:
    ``{"status": "success", "body": ...}``

    Rejection shape:
    ``{"status": "error", "error_details": {...}}``
    """

    status: str
    body: Optional[T] = None
    details: Optional[ApiRejectedDetails] = None

    @classmethod
    def from_dict(cls, data: dict) -> "ApiResponse[T]":
        status = str(data.get("status", ""))
        if status == "success":
            return cls(status=status, body=data.get("body"))
        if status == "error":
            return cls(
                status=status,
                details=ApiRejectedDetails.from_dict(data.get("error_details") or {}),
            )
        raise ValueError(f"Unsupported ApiResponse status: {status}")
